README.md
NFT Staking Platform for VexedApes on Near Protocol
Kokumo Version

changelog.md
# Changelog

## 1.0.10 several small standards updates, see #8, #9, #10, #12, #13, #15, #16, #17

- several optional args for nft_transfer_payout are not optional anymore
- nft_transfer requires approval_id
odd since if you're owner you should be able to transfer without approval_id - just pass 0 if you're owner, see tests
- check diff for changes to tests if you rolled your own: https://github.com/near-apps/nft-market/commit/ba2b5ce07043061059a5482e815433101d4c455f
- check diff for changes to market, max_len_payout is required parameter now, your market may vary in how many payouts it allows

## 1.0.9 remove option on u64 for nft_transfer[_payout,_call]

- Makes NFT match standard closely, even though calls from Market/Client unaffected e.g. they could send approval_id or not

## 1.0.8 fix u64 vs. U64 for approval ids

- Makes NFT and Market contract match standard
- Re-deployment is non-breaking to existing NFT contracts, even if approvals are already added
- Recommeneded: Markets should remove all sales prior to migrating since since approval_id type is changed

## 1.0.7 fix breaking change to sale_views.rs

- Add storage_minimum_balance to nft-staking 

## 1.0.7 fix breaking change to sale_views.rs

- Add storage_minimum_balance to nft-staking

## 1.0.6 fix breaking change to sale_views.rs

- Bug in nft-staking storage_withdraw thanks @BenKurrek
- Add storage standard interface to nft-staking

## 1.0.5 fix breaking change to sale_views.rs

- Bug in get_sales_by_nft_contract_id

## 1.0.4 support wallet NFT pagination

- Update contracts to make limit: number (vs string)
- Update UI, api-helper was also released to accomodate this change

## 1.0.3 bug fix

- Wallet not connected

## 1.0.2 refactor and bug fix

- Refactor views and actions
- Fix for sale_conditions rename in market contract

## 1.0.1 gh ci

- Passing CI for Node 15.x

## 1.0.0 initial release

- Feature freeze
- API interface client passed args freeze


test_staking_contract.js
// Welcome to the Mass Key Deletion recipe.

// This tool allows you to
// 1. Delete all your functionCall Access Keys
// 2. Delete all but one specified Full Access Key
// 3. Delete all Full Access Keys and Lock an Account

/// STEP 1 Install near-api-js
// npm init (in directory where you stored this script)
// npm i near-api-js

const nearAPI = require("near-api-js"); // imports near api js
const { parseNearAmount } = require("near-api-js/lib/utils/format");

// Standard setup to connect to NEAR While using Node
const { keyStores, connect } = nearAPI;
const homedir = require("os").homedir();
const CREDENTIALS_DIR = ".near-credentials";
const credentialsPath = require("path").join(homedir, CREDENTIALS_DIR);
const keyStore = new keyStores.UnencryptedFileSystemKeyStore(credentialsPath);
let config;

// STEP 2 Choose your configuration.
// set this variable to either "testnet" or "mainnet"
// if you haven't used this before use testnet to experiment so you don't lose real tokens by deleting all your access keys
const configSetting = "mainnet";

const GAS_FOR_NFT_APPROVE = "20000000000000";
const GAS_FOR_RESOLVE_TRANSFER = "10000000000000";
const MAX_GAS = "300000000000000";
const GAS_FOR_FT_TRANSFER = "4000000000000";
const DEPOSIT = "450000000000000000000";

// setting configuration based on input
switch (configSetting) {
  case "mainnet":
    config = {
      networkId: "mainnet",
      keyStore, // optional if not signing transactions
      nodeUrl: "https://rpc.mainnet.near.org",
      walletUrl: "https://wallet.near.org",
      helperUrl: "https://helper.mainnet.near.org",
      explorerUrl: "https://explorer.mainnet.near.org",
    };
    console.log("configuration set to mainnet ");

    break;

  case "testnet":
    config = {
      networkId: "testnet",
      keyStore, // optional if not signing transactions
      nodeUrl: "https://rpc.testnet.near.org",
      walletUrl: "https://wallet.testnet.near.org",
      helperUrl: "https://helper.testnet.near.org",
      explorerUrl: "https://explorer.testnet.near.org",
    };
    console.log("configuration set to testnet ");
    break;
  default:
    console.log(`please choose a configuration `);
}

const STAKING_CONTRACT_ID = "stakingancwallet2.near";
const TOKEN_CONTRACT_ID = "anctokenwallet.near";
const NFT_CONTRACT_ID = "apenearclub.near";

const Test = async () => {
  //Load Your Account
  const near = await connect(config);

  // STEP 4 enter your mainnet or testnet account name here!
  const account = await near.account("stakingancwallet2.near");
  const tokenContractAccount = await near.account(TOKEN_CONTRACT_ID);

  let result;
  result = await account.viewFunction(
    TOKEN_CONTRACT_ID,
    "ft_balance_of",
    {
      account_id: account.accountId,
    }
  );
  console.log("stakingancwallet2.near ANCAmount:", result);

  result = await account.viewFunction(
    STAKING_CONTRACT_ID,
    "get_claim_amount",
    {
      account_id: account.accountId,
    }
  );
  console.log("ClaimAmout:", result);

  // transfer anc token to staking contract
  // result = await tokenContractAccount.functionCall({
  //   contractId: TOKEN_CONTRACT_ID,
  //   methodName: "ft_transfer",
  //   args: {
  //     receiver_id: STAKING_CONTRACT_ID,
  //     amount: parseNearAmount("3000000"),
  //     msg: JSON.stringify({ msg: "Stake to Platform" })
  //   },
  //   gas: GAS_FOR_FT_TRANSFER,
  //   attachedDeposit: "1",
  // })

  // STAKING
  // result = await account.functionCall({
  //   contractId: NFT_CONTRACT_ID,
  //   methodName: "nft_approve",
  //   args: {
  //     token_id: "QmVmz2KGaWW9qWvzajHSGNjpx1odxaXBaCGeaNwd8JsFyo",
  //     account_id: STAKING_CONTRACT_ID,
  //     msg: JSON.stringify({ staking_status: "Stake to Platform" })
  //   },
  //   gas: MAX_GAS,
  //   attachedDeposit: DEPOSIT,
  // });

  // CLAIMING
  // result = await account.functionCall({
  //   contractId: STAKING_CONTRACT_ID,
  //   methodName: "claim_reward",
  //   args: {
  //   },
  //   gas: MAX_GAS,
  //   attachedDeposit: "1",
  // });
  // console.log(result);

  // UNSTAKING
  // result = await account.functionCall({
  //   contractId: STAKING_CONTRACT_ID,
  //   methodName: "unstake",
  //   args: {
  //     token_id: "QmVmz2KGaWW9qWvzajHSGNjpx1odxaXBaCGeaNwd8JsFyo" 
  //   },
  //   gas: MAX_GAS,
  //   attachedDeposit: "1",
  // });
  // console.log(result);
};

Test();


utils/deploy.js
const fs = require("fs");
const testUtils = require("../test/test-utils");
const { GAS } = getConfig();
const { createOrInitAccount } = testUtils;

/// Manual deployment

// 'vadim-nfts.testnet'
// near create_account market.name.testnet --masterAccount=name.testnet --initial-balance 50
// near deploy --accountId=market.name.testnet --wasmFile=./out/market.wasm
// near deploy --accountId=name.testnet --wasmFile=./out/main.wasm
// near call market.name.testnet new '{"owner_id": "name.testnet", "ft_token_ids": []}' --accountId=name.testnet
// near call name.testnet new '{"owner_id":"name.testnet","metadata":{"spec":"name","name":"NFT","symbol":"NFT"},"supply_cap_by_type":{"test": "1000000"}}' --accountId=name.testnet

const deployAccount = await createOrInitAccount(
  process.env.DEPLOY_ACCOUNT_ID,
  process.env.DEPLOY_ACCOUNT_SECRET
);
const contractBytes = fs.readFileSync("./out/main.wasm");
console.log("\n\n Deploying NFT Contract \n\n");
const newArgs = {
  owner_id: contractAccount.accountId,
  metadata: {
    spec: "nft-1",
    name: "vadim-nfts",
    symbol: "VNFT",
  },
};

const actions = [
  deployContract(contractBytes),
  functionCall("new", newArgs, GAS),
];
await deployAccount.signAndSendTransaction(stableId, actions);

/// TODO clean up market deployment

/// create or get market account and deploy market.wasm (if not already deployed)
marketAccount = await createOrInitAccount(marketId, GUESTS_ACCOUNT_SECRET);
const marketAccountState = await marketAccount.state();
console.log("\n\nstate:", marketAccountState, "\n\n");
if (marketAccountState.code_hash === "11111111111111111111111111111111") {
  const marketContractBytes = fs.readFileSync("./out/market.wasm");
  console.log(
    "\n\n deploying marketAccount contractBytes:",
    marketContractBytes.length,
    "\n\n"
  );
  const newMarketArgs = {
    owner_id: contractId,
    ft_token_ids,
  };
  const actions = [
    deployContract(marketContractBytes),
    functionCall("new", newMarketArgs, GAS),
  ];
  await marketAccount.signAndSendTransaction(marketId, actions);

  /// NOTE market must register for all ft_token_ids it wishes to use (e.g. use this loop for standard fts)
  ft_token_ids.forEach(async (ft_token_id) => {
    const deposit = await marketAccount.viewFunction(
      ft_token_id,
      "storage_minimum_balance"
    );
    await marketAccount.functionCall(
      ft_token_id,
      "storage_deposit",
      {},
      GAS,
      deposit
    );
  });
}
// get all supported tokens as array
const supportedTokens = await marketAccount.viewFunction(
  marketId,
  "supported_ft_token_ids"
);
console.log(
  "\n\n market supports these fungible tokens:",
  supportedTokens,
  "\n\n"
);

// should be [false], just testing api
const added = await contractAccount.functionCall(
  marketId,
  "add_ft_token_ids",
  { ft_token_ids },
  GAS
);
console.log("\n\n added these tokens", supportedTokens, added, "\n\n");

/// find out how much needed for market storage
storageMarket = await contractAccount.viewFunction(marketId, "storage_amount");
console.log("\n\n storageMarket:", storageMarket, "\n\n");


for_token_airdrop.js
const nearAPI = require("near-api-js"); // imports near api js
const { parseNearAmount } = require("near-api-js/lib/utils/format");

// Standard setup to connect to NEAR While using Node
const { keyStores, connect } = nearAPI;
const homedir = require("os").homedir();
const CREDENTIALS_DIR = ".near-credentials";
const credentialsPath = require("path").join(homedir, CREDENTIALS_DIR);
const keyStore = new keyStores.UnencryptedFileSystemKeyStore(credentialsPath);
let config;

// STEP 2 Choose your configuration.
// set this variable to either "testnet" or "mainnet"
// if you haven't used this before use testnet to experiment so you don't lose real tokens by deleting all your access keys
const configSetting = "mainnet";

const GAS_FOR_NFT_APPROVE = "20000000000000";
const GAS_FOR_RESOLVE_TRANSFER = "10000000000000";
const GAS_FOR_FT_TRANSFER= "4000000000000";
const MAX_GAS = "300000000000000";
const DEPOSIT = "450000000000000000000";
const ONE_YOCTO = "1";
// setting configuration based on input
switch (configSetting) {
  case "mainnet":
    config = {
      networkId: "mainnet",
      keyStore, // optional if not signing transactions
      nodeUrl: "https://rpc.mainnet.near.org",
      walletUrl: "https://wallet.near.org",
      helperUrl: "https://helper.mainnet.near.org",
      explorerUrl: "https://explorer.mainnet.near.org",
    };
    console.log("configuration set to mainnet ");

    break;

  case "testnet":
    config = {
      networkId: "testnet",
      keyStore, // optional if not signing transactions
      nodeUrl: "https://rpc.testnet.near.org",
      walletUrl: "https://wallet.testnet.near.org",
      helperUrl: "https://helper.testnet.near.org",
      explorerUrl: "https://explorer.testnet.near.org",
    };
    console.log("configuration set to testnet ");
    break;
  default:
    console.log(`please choose a configuration `);
}

const STAKING_CONTRACT_ID = "kokumokongz_staking_wallet.near";
const TOKEN_CONTRACT_ID = "kokumokongz_token_wallet.near";
const NFT_CONTRACT_ID = "banc.neartopia.near";

const AirDrop = async () => {
  //Load Your Account
  const near = await connect(config);

  // STEP 4 enter your mainnet or testnet account name here!
  const account = await near.account(TOKEN_CONTRACT_ID);

  let result;
  result = await account.getAccessKeys();
  console.log(result);

  let airdrop = [
    {
      name: "veejayrex.near",
      amount: "200"
    },
    {
      name: "hessmart.near",
      amount: "200"
    },
    {
      name: "newein.near",
      amount: "100"
    },
    {
      name: "a5e34454d4262f68c2a6613b76f6ecfe1f00cd3af08698e1ae4b5247d1d6e9dd",
      amount: "200"
    },
    {
      name: "moffelvdbuurt.near",
      amount: "500"
    },
    {
      name: "khanamjad096.near",
      amount: "100"
    },
    {
      name: "aldi100.near",
      amount: "100"
    },
    {
      name: "r76.near",
      amount: "200"
    },
    {
      name: "codedforum.near",
      amount: "100"
    },
    {
      name: "onium.near",
      amount: "100"
    },
  ]
  for(let i=0; i<airdrop.length; i++){
    const resultSend = await account.functionCall({
      contractId: TOKEN_CONTRACT_ID,
      methodName: "ft_transfer",
      args: {
        receiver_id: airdrop[i].name,
        amount: parseNearAmount(airdrop[i].amount),
        memo: JSON.stringify({ msg: "Airdrop to NFT Holder" })
      },
      gas: GAS_FOR_FT_TRANSFER,
      attachedDeposit: ONE_YOCTO,
    });
  }

  // result = await account.viewFunction(
  //   NFT_CONTRACT_ID,
  //   "nft_tokens",
  //   {
  //     from_index: "580",
  //     limit: 15,
  //   }
  // ); 

  // function isEmpty(obj) {
  //   for(var prop in obj) {
  //     if(Object.prototype.hasOwnProperty.call(obj, prop)) {
  //       return false;
  //     }
  //   }
  
  //   return JSON.stringify(obj) === JSON.stringify({});
  // }

  // for(let i=0; i<result.length; i++){
  //   if(isEmpty(result[i].approved_account_ids) && result[i].owner_id != "vexpremint.near" )
  //   {
  //     console.log(result[i].token_id, ":", result[i].owner_id);
  //     try{
  //       const resultSend = await account.functionCall({
  //         contractId: TOKEN_CONTRACT_ID,
  //         methodName: "ft_transfer",
  //         args: {
  //           receiver_id: result[i].owner_id,
  //           amount: parseNearAmount("100"),
  //           memo: JSON.stringify({ msg: "Airdrop to NFT Holder" })
  //         },
  //         gas: GAS_FOR_FT_TRANSFER,
  //         attachedDeposit: ONE_YOCTO,
  //       });
  //     } catch (error){
  //       console.log(error);
  //     }
  //   }
  // }

  // STAKING
  // result = await account.functionCall({
  //   contractId: NFT_CONTRACT_ID,
  //   methodName: "nft_approve",
  //   args: {
  //     token_id: "QmVmz2KGaWW9qWvzajHSGNjpx1odxaXBaCGeaNwd8JsFyo",
  //     account_id: STAKING_CONTRACT_ID,
  //     msg: JSON.stringify({ staking_status: "Stake to Platform" })
  //   },
  //   gas: MAX_GAS,
  //   attachedDeposit: DEPOSIT,
  // });

  // CLAIMING
  // result = await account.functionCall({
  //   contractId: STAKING_CONTRACT_ID,
  //   methodName: "claim_reward",
  //   args: {
  //   },
  //   gas: MAX_GAS,
  //   attachedDeposit: "1",
  // });
  // console.log(result);

  // UNSTAKING
  // result = await account.functionCall({
  //   contractId: STAKING_CONTRACT_ID,
  //   methodName: "unstake",
  //   args: {
  //     token_id: "QmVmz2KGaWW9qWvzajHSGNjpx1odxaXBaCGeaNwd8JsFyo" 
  //   },
  //   gas: MAX_GAS,
  //   attachedDeposit: "1",
  // });
  // console.log(result);
};

AirDrop();
